### DISCUSSION TITLE
Should we use the term "deprecated"?

### DISCUSSION THREAD

**Discussion by ddbeck** [mdn/mdn-community#566] [Content]
Last week, I had a lengthy conversation with @wbamberg, @Elchi3, @sideshowbarker, and others on some trouble with using the term "deprecated" on MDN and in BCD. I'm mostly writing this down to get it out of my head, but if there are others who are interested in this discussion, please point them to this discussion.

The questions that came out of that conversation came out to three parts:

- What do we intend for "deprecated" to mean on MDN?
- Is our use of "deprecated" understood by readers in the same way we intend it?
- Should we continue to use "deprecated" or consider adopting alternatives?

I've got some additional background for each of these questions in the sections that follow.

## What do we intend for "deprecated" to mean on MDN?

Since BCD drives a lot of the deprecated text on MDN (e.g., the tooltips on deprecation in compat tables), BCD's definition of deprecated happened to spur this whole discussion.

> If deprecated is `true`, then the feature is no longer recommended. It might be removed in the future or might only be kept for compatibility purposes. Avoid using this functionality.

An interesting question here is: does BCD agree with other practices (e.g., adding deprecation banners) on MDN? As an author, do you typically use "deprecated" to mean anything else?

## Is our use of "deprecated" understood by readers in the same way we intend it?

In https://github.com/mdn/browser-compat-data/issues/10490, a contributor mentioned this Stack Overflow question: [Is the JavaScript keyword `with` really deprecated?](https://stackoverflow.com/questions/48810592/is-the-javascript-keyword-with-really-deprecated). The discussion there highlights some of the conflict: TC39 uses "deprecated" one way, MDN another, an argument ensues. I don't know if we have other (good) ways to find out what readers think "deprecated" means.

To get a sense of the broader use of "deprecated", I asked in the Write the Docs Slack about how other tech writers are using "deprecated" (see [my full notes](https://gist.github.com/ddbeck/96650af1ed5228668f173e07e5417909#file-summary-md) for a complete breakdown of the discussion). The biggest takeaway there is that other authors often use "deprecated" to mean that something is on the road to _removal_, even if it never arrives at that destination. My conclusion is that MDN is doing something slightly unorthodox, but I'm curious if there are other perspectives on this.

## Should we continue to use "deprecated" as we've been doing or consider adopting alternatives?

The Write the Docs discussion highlighted a key thing, which is probably of interest regardless of how we use "deprecated": when something is discouraged from use we ought to direct users to alternatives.

Apart from that, there are probably other words we could use in place of deprecated, if we think we're using deprecated wrongly. We'd need to come up with some alternatives and have a definition to go with it.

## Related issues

https://github.com/mdn/yari/discussions/2905 - discusses using front matter to set deprecation banners

**Comment by sideshowbarker** [comment]
> Apart from that, there are probably other words we could use in place of deprecated, if we think we're using deprecated wrongly. We'd need to come up with some alternatives and have a definition to go with it.

I think the word that best fits with what we’re wanting to convey is _“discouraged”_.

_“discouraged”_ is good because it lacks the baggage and formal connotations of _“deprecated“_ and of _“obsolete”_ too.

There are cases when specs have need to normatively assert formal status of a feature — and that spec need is met by _“deprecated“_ and of _“obsolete”_. But for MDN, we have no need to normatively (re)assert formal status of anything. Instead, we only have need to convey a less formal assessment of whether particular features are *“not generally recommended”* or *“known not to be a best practice”*.

So *“discouraged”* is a good concise word to convey *“not generally recommended”* or *“known not to be a best practice”*.

As far as relevant discussion elsewhere: TC39, for stuff that’s part of the language but that’s generally discouraged — in particular, much of the stuff in [Annex B of the ES spec](https://tc39.es/ecma262/multipage/additional-ecmascript-features-for-web-browsers.html#sec-additional-ecmascript-features-for-web-browsers) — has been internally using the word *“icky”*…

> Is our use of "deprecated" understood by readers in the same way we intend it?

I think we have evidence it’s confusing to some readers, and that other readers object to some cases where we’re using it.

But I think *“discouraged”* would at least not be as confusing, and at least not as objectionable.

To help put it in terms of what developers are already familiar with: I think most developers understand what [“code smell” ](https://en.wikipedia.org/wiki/Code_smell) means — and if they aren’t familiar with that term already, they get the interference — and I’m reminded of the Zappa quote about “Jazz isn't dead. It just smells funny”.

What we need for MDN is something to flag cases of _“This feature may not be dead yet, and may not be formally recognized as even being on its way to becoming dead, and in fact may never die — but… it’s kinda icky and it kinda smells funny”_.

> Should we continue to use "deprecated" or consider adopting alternatives?

I believe we have a clear need to continue flagging the stuff we currently have flagged as deprecated — but I think we just need to come up with a better word for conveying it to developers. And short of going all the way to “icky” or ”smells funny”, I think _“discouraged”_ conveys well what we actually are wanting to convey.

**Reply by wbamberg** [reply]
+1 for "discouraged"

**Reply by jpmedley** [reply]
I like 'discouraged' but wonder if we should be concerned that it suggests MDN is expressing an opinion rather than making a statement of fact: i.e. that a particular feature is being phased out.

I can think of other ways that 'deprecated' is overloaded other than those mentioned above. I won't elaborate because I can only speculate on whether and how that causes confusion.

**Reply by hamishwillee** [reply]
^^^ Yes. When I see "discouraged" I assume that I should not use the thing - but I do not assume a particular reason. Could be a formal specification deprecation ("on the pathway to removal") or could just be "not a good idea according to us". Usually I'd hope for additional context to be provided for something that it discouraged: discouraged because ....

By contrast I see deprecation as having some formal meaning, and spec driven. I think globally replacing the word a loss of useful information. That might be OK, but it should be a decision we deliberately make.

Had a case with @ericwbailey recently around something to do with Aria where the feature was discouraged but not deprecated.  In that case we replaced the deprecated macro with a warning note that captured this "discourage use".
That was the first time this distinction required me to take an action.

EDITED: In other words, I'm not so worried about people's issues with the meaning of deprecation - whether or not they think it means "on the way to removal" they will understand it means "you should not use this". I'm more worried they might think that discouraged is a personal opinion of MDN.

**Reply by ericwbailey** [reply]
To add a little more color to this circumstance, the HTML element in question is considered actively harmful to assistive technology, as it negated roles for a common navigation method.

I think "discouraged" is a good term, but I'm also wondering if we need a term for something that has this strong of an adverse effect. It feels separate from something that can still technically function due to compat, but isn't considered best practice anymore. 

I feel many developers will grab something that technically works and not read the fine print, so that's my main concern.

**Reply by sideshowbarker** [reply]
> ^^^ Yes. When I see "discouraged" I assume that I should not use the thing - but I do not assume a particular reason. Could be a formal specification deprecation ("on the pathway to removal") or could just be "not a good idea according to us".

I’m not aware of any actual "not a good idea according to us" cases — that is, where something in MDN has been arbitrarily flagged a feature just because somebody personally felt it wasn’t good.

I think there’s always a good reason for some feature having been flagged — and in many cases it’s not because a specification has formally marked the feature has deprecated.

Part of the reason that “deprecated” is messy and complicated is that specs don’t consistently use the term to mean the same thing — or don’t even use it at all.

And that creates confusion for developers.

But rather than just assert that, I guess it’s more helpful to look at concrete cases in specific specs. So here are two concrete cases — one from the ECMAScript spec and one from the HTML spec.

---

### 1. ECMAScript spec

Consider https://github.com/mdn/content/issues/5509. The OP for that issue asserted that MDN shouldn’t be flagging `Object.prototype.__proto__` as deprecated — because it’s in the ES spec at https://tc39.es/ecma262/multipage/additional-ecmascript-features-for-web-browsers.html#sec-additional-properties-of-the-object.prototype-object and not clearly marked as deprecated there.

And they are right. The ES spec doesn’t clearly mark it as deprecated — nor, in that section where it’s defined, have any language warning developers not to use it.

For a developer reading the spec to start to understand more of the context, they’d need to think to go look at https://tc39.es/ecma262/multipage/additional-ecmascript-features-for-web-browsers.html#sec-additional-ecmascript-features-for-web-browsers — the intro for Annex B of the spec (the spec section that feature is in) — and read the note at the beginning of that, and notice the parts that say:

> All of the language features and behaviours specified in this annex have one or more undesirable characteristics and in the absence of legacy usage would be removed from this specification.
>
> Programmers should not use or assume the existence of these features and behaviours when writing new ECMAScript code. ECMAScript implementations are discouraged from implementing these features

…but nowhere even there does the word “deprecated” occur.

In fact nowhere in [the entirety of the ES spec](https://tc39.es/ecma262/) does it flag anything as “deprecated” — nor “obsolete”, nor any other metonym.

And even in a case such as the `Object.prototype.__proto__` it’s not clear-cut. To get some sense of the subtleties, you’d need to do something like reading [the relevant TC39 meeting notes](https://github.com/tc39/notes/blob/master/meetings/2020-09/sept-21.md#conclusionresolution-1), and see stuff like this:

* The proto syntax will be required and not marked as discouraged
* The proto accessor will be optional and marked as discouraged.
* The defineGetter, defineSetter, lookupGetter, and lookupSetter will be optional and discouraged.

OK, so in our case, the https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/proto is documenting the accessor, thus it’s appropriate for us to flag it.

But then we’re back to the fact that we’re currently flagging it as “Deprecated” despite the fact that any developer can go in read the spec and see that it’s not formally marked as deprecated. So the closest we can get for this really is “Discouraged”.

---

### 2. HTML spec

The HTML spec also never flags as anything “deprecated”. Instead it uses the stronger term “obsolete” (which by policy we’re no longer using in MDN…), in combination with “(non)conforming”. And the [Obsolete features](https://html.spec.whatwg.org/multipage/obsolete.html#obsolete) section of the spec subclasses those obsolete features into “obsolete but conforming” and “obsolete and non-conforming”.

But anyway, that’s actually all very clearly and precisely defined. For MDN purposes, it’s clearly appropriate for us to flag both the “obsolete but conforming” and “obsolete and non-conforming” features as “deprecated”.

However, the HTML spec contains other cases which are not nearly as clear. For example, consider `NavigatorID` https://html.spec.whatwg.org/multipage/system-state.html#client-identification.

Among the `NavigatorID` members that get exposed are:

* navigator.appCodeName, which is required to always return `Mozilla`
* navigator.appName, which is required to always return `Netscape`
* navigator.product, which is required to always return `Gecko`

In other words, none of those features are actually useful for anything in practice — because they always return the same hardcoded string, regardless of which browser a user is running — and it’s appropriate to warn developers not to try to use them for anything. They’re literally useless.

Yet, the HTML spec doesn’t put them into the “Obsolete features” section, nor does the HTML spec flag them in any way which makes it clear to developers that they shouldn’t be used.

So for MDN, what should we do with them? As with the ES spec cases, flagging those as “deprecated” conflicts with how the spec itself categorizes them. And in particular, for those cases, it’s inaccurate for us to show warning text that includes:

> Though some browsers might still support it, it may have already been removed from the relevant web standards… Be aware that this feature may cease to work at any time

…because `navigator.appCodeName` will never cease to work, and will never be removed from the HTML spec. Instead, it will be kept in the spec forever, with a precisely defined requirement about what it must do. It’s just that the spec itself actually makes it, be definition, a useless feature that developers should avoid doing anything with.

So that seems like another case where the closest we can really get is “Discouraged” or some such.

---

> Had a case with @ericwbailey recently around something to do with Aria where the feature was discouraged but not deprecated. In that case we replaced the deprecated macro with a warning note that captured this "discourage use".

That seems like another great example of where having a Discouraged flag would solve a problem.

And to be clear about something: I think the Discouraged flag could cover both things we need:

* cases like that ARIA case and the ES spec and HTML spec cases I described here; and also
* cases where a spec has actually clearly flagged something as “deprecated” or “obsolete”

In other words, we can use “discouraged” in the broader sense that also includes “deprecated” and “obsolete”.

> Usually I'd hope for additional context to be provided for something that it discouraged: discouraged because ....

I think in most cases we’re already doing that carefully and consistently — for example, the ARIA case you mentioned, or something like https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/proto (which has more than 130 words of specific warning text), or something like https://developer.mozilla.org/en-US/docs/Web/API/NavigatorID/appName (which has text saying _This property is kept only for compatibility purposes_ and _Do not rely on this property to return a real browser name. All browsers return "Netscape" as the value of this property_), and so on.

> I'm more worried they might think that discouraged is a personal opinion of MDN.

For both sets of cases we’re considering — both the “there’s a spec which clearly marks this feature as deprecated”, and the gray-area cases discussed here — I cannot recall any feature we have flagged in an MDN article without some specific text explaining to developers why they should avoid that feature.

That is, I can’t think of any cases where it’s not already clear that the article isn’t just asserting something that’s a personal opinion of MDN.

That said, I’m not convinced there aren’t any such cases of articles that lack a clear explanation for a feature being flagged — in fact I’m nearly certain we must have some (and I wouldn’t be surprised if today I actually now come across one…). But I think solution to that is to deal with them case by case as we notice them, and make sure they get updated with good explanations.

But the core problem remains that we’re currently flagging many of the gray-area cases as Deprecated —  for example, the `Object.prototype.__proto__` and the `navigator.appName` cases — when in fact they’re not actually deprecated. And so it would be useful to have some term we can consistently use to flag those (rather than instead needing to do it solely with ad-hoc language, as happened with that ARIA case).

**Reply by ddbeck** [reply]
It sounds to me like there's a general good feeling toward using "discouraged" to mean "in some way indicated by an authority—a standards body, usually—that the discouraged thing is officially disfavored" and a consensus that we shouldn't editorialize, if we can help it. Perhaps we could experiment with "discouraged" in the scope of BCD (e.g., revise the compat schema to define the current `"deprecated"` key this way and update the compat tables on MDN to use "discouraged" in the tool tip),

Also, good news on the ECMAScript front, specifically: [it looks like they'll be adopting some "legacy" terminology to formalize "icky"](https://github.com/tc39/ecma262/pull/2441#issuecomment-866968681). I learned that as a consequence of my PR to get a warning applied to `with`. I'll be updating my PR there to use the legacy modifier, though I imagine it will be sometime before it's considered by the committee and, hopefully, accepted.

**Reply by hamishwillee** [reply]
@sideshowbarker @ddbeck FYI this came up again today here: https://github.com/nodejs/node/pull/39581#discussion_r680576012

Has this progressed at all? 
Discouraged in the tooltip would help a bit. Also perhaps making the banner say discouraged and explain what we mean.
But what I think might really help would be an optional way to specify WHY something is discouraged. This could have a couple of standard reasons - "dropped from spec", "inconsistent browser support", whatever.

**Comment by sideshowbarker** [comment]
https://github.com/mdn/content/issues/4866 relates to another grey-area feature: `node.isSameNode()`.

The DOM spec doesn’t mark it as deprecated — all that is does say is a *comment* in the IDL:
https://dom.spec.whatwg.org/#ref-for-dom-node-issamenode
```webidl
boolean isSameNode(Node? otherNode); // legacy alias of ===
```
The *implication* of “legacy alias of ===” is “You shouldn’t use `isSameNode()` but should just use `===` instead. But the spec doesn’t say that explicitly — and to most people reading the spec, it’s probably not even going to be clear that’s what the spec is implying.

**Reply by ddbeck** [reply]
Like my `with` statement proposal, maybe that's another place where we ought to go upstream and ask for changes to the specification?

**Reply by sideshowbarker** [reply]
> Like my `with` statement proposal, maybe that's another place where we ought to go upstream and ask for changes to the specification?

Yes, I’ll bring it up with Anne

**Comment by sideshowbarker** [comment]
For the record here: For the ES spec, TC39 adopted and documented the use of the special word/flag *Legacy* — for flagging cases in the spec of features that they want to warn developers away from.

https://tc39.es/ecma262/multipage/conformance.html#sec-conformance
> All of the language features and behaviours specified within Legacy subclauses have one or more undesirable characteristics. However, their continued usage in existing applications prevents their removal from this specification. These features are not considered part of the core ECMAScript language. Programmers should not use or assume the existence of these features and behaviours when writing new ECMAScript code.

**Reply by jpmedley** [reply]
_Legacy_ sounds like something that will stick around. We need a term that sounds like it's going away.

**Reply by hamishwillee** [reply]
It's a step in a good direction for the spec to have that definition. Pity it doesn't spell out more clearly: "prevents their removal from this specification _at this time_".

For MDN I prefer "discouraged" if we are keeping a single un-nuanced "we don't think you should use this".

**Comment by jpmedley** [comment]
When a language discussion gets nowhere, I think it's helpful to look again at the thing it's trying to describe. Having written Chrome's own [deprecations and removals column](https://developer.chrome.com/tags/removals/) for the last six years, I think I can make a few generalizations.

First, as a web developer, I don't care about a spec deprecation until a browser implements it. I don't know what other browsers do, but when Chrome implements a deprecation, we either build its replacement first, or we look closely at how much something is used in the wild. So the sequence in Chrome looks like this:

Spec deprecation > [Replacement implementation] > Chrome deprecation(includes console warning) > Chrome removal

Question: can't we just solve this with adjectives? "Spec deprecation" "Browser deprecation".

**Comment by hamishwillee** [comment]
> Question: can't we just solve this with adjectives? "Spec deprecation" "Browser deprecation".

I wish we could. MDN uses deprecated to indicate things that might be neither, as per the discussion further up: https://github.com/mdn/content/discussions/5549#discussioncomment-837662

**Reply by jpmedley** [reply]
There are multiple references to MDN's use of deprecated, but I don't see it described explicitly. Is it written down officially somewhere? If not, how can we be sure that Mozillian's all have the same definitions in their heads?

**Reply by hamishwillee** [reply]
@jpmedley AFAIK The reason this discussion was started was to agree and define that common understanding of the term "deprecated", or replace it with another term that is closer to what we mean. There is an implied assumption that having a single term for this concept is a good thing, and that it can't be "deprecated" because that implies "on the path to being removed" - which not all cases are. I have not questioned those assumptions.

The closest this got to converging was the term "discouraged" as  a single catch-all for "don't use this for a whole bunch or possible reasons".

**Reply by jpmedley** [reply]
>it can't be "deprecated" because that implies "on the path to being removed" - which not all cases are.

I'm not seeing what those cases actually are. What are they?

**Reply by hamishwillee** [reply]
The only case I know of was https://github.com/mdn/content/discussions/5549#discussioncomment-837662 - I'm assuming it's a bigger problem though than just one case, or we'd deal with as an exception.

**Reply by jpmedley** [reply]
I don't think that needs a word of its own that the reader may or may not interpret correctly. I think it needs a big fat blurb explaining exactly what the problem is. If that's the only alternative use case, then it shouldn't soak up a word that has broader use elsewhere, particularly since "deprecated" may not always mean "don't use now". In fact, in some cases it may say just the opposite. There are cases where for interoperability, a developer may need to write code that uses a newer, preferred API if it's available, and falls back to an earlier deprecated one when it is not. That's a long way of saying that "deprecated" sometimes, at least in the case of spec deprecations, means "use carefully" or "use sparingly."

**Reply by hamishwillee** [reply]
@jpmedley If you read the first post, both individuals and specs are using deprecated to mean different things.

@ddbeck You might want to loop back into the conversation. I don't want to be the person "defending" various options here when I don't actually have any strong opinion on this (other than we need clear common understanding of whatever terminology we use). I'm sure you have a much firmer grip on the relative options.

**Reply by jpmedley** [reply]
>If you read the first post, both individuals and specs are using deprecated to mean different things.

I'm for specifics and I'm not getting them. The original statement, which I keep getting referred back to, has no information beyond an assertion that something is so.

**Reply by ddbeck** [reply]
> The original statement, which I keep getting referred back to, has no information beyond an assertion that something is so.

@jpmedley is the earlier TC39 "legacy" example not sufficient? It's a clear case where the spec is saying "[don't use this](https://github.com/tc39/ecma262/pull/2441/files)" while also maintaining that the legacy features will never be removed from the spec (as part of a commitment to backwards compatibility).

More broadly, I haven't had time to act on this in more substantial ways, though I think it's important and I want it to become a part of my planned efforts. That said, I'm still dealing with some onboarding stuff and some urgent matters, so I haven't been in a position to make a concrete plan yet.

Responding to the thread in general, I think we need _both_ "fat blurbs" and a schematic way to refer to things which are in the category of things which are deprecated, obsolete, legacy, and what have you. "Deprecated" is plainly misunderstood (or sometimes rejected) by readers (see: the last few questions about this within BCD's own issues). We use that term in lots of places and we really shouldn't, due to reader confusion. We also use that term _as_ the schematic (as in BCD), where we shouldn't. We'll also need to accomodate changes within the category that doesn't require us to rewrite every time a convention changes

For example, before July, TC39 never had a single deprecation formalism, but now it does. That probably didn't need to matter to the way we represented it to readers or the way we captured it in data, except that the term we used provoked annoyance or disagreement.

I think "discouraged" wraps up a lot of this stuff schematically, but it's not an answer for the other pieces (I've got a more detailed take on how to do this, which [I've written up on my own site](https://ddbeck.com/deprecated-pattern/). It's not a propsal to carry out the work, however.). We'll need a comprehensive effort to deal with this stuff.

**Comment by jpmedley** [comment]
>@jpmedley is the earlier TC39 "legacy" example not sufficient? It's a clear case where the spec is saying "don't use this" while also maintaining that the legacy features will never be removed from the spec (as part of a commitment to backwards compatibility).

I can live with 'legacy' for TC39 features because it's in the spec. I object to cases where it's not in the source. For example, Chrome uses the word 'deprecated' to refer to features slated for removal at a later date. To not mirror that term in BCD or on MDN is to introduce a possible source of developer confusion. I can imagine someone asking what's the difference between a Chrome deprecated feature and a Chrome legacy feature and, not knowing it's MDN's term, searching in vain for an explanation on a Google site.

**Comment by ericwbailey** [comment]
@ddbeck Your [PR for removing HTML Imports](https://github.com/mdn/content/pull/9292) reminded me of this discussion. 

I understand removing this content due to lack of current and future support, but also am wondering if there is a need to document this content as having been a thing for the historical record? I would imagine it would have all the necessary surrounding content to warn about support.

**Reply by ddbeck** [reply]
@ericwbailey narrowly, I think reference documentation should be relevant to contemporary developers and we shouldn't make historical preservation an aim. Readers skim more than they actually read, so it's risky to keep full documentation around for things which are plainly unusable, even with warnings. That said, sometimes you need coverage on historic features to explain other, contemporary features ([an example with `Accept-Charset`](https://github.com/mdn/content/pull/4250)).

More broadly, perhaps there's some value in a sort of feature graveyard document that very briefly acknowledges the historic existence of features or proposals and refers readers to contemporary alternatives. That could be good to write (and a better target for redirects), but it's also very difficult to prioritize, for hopefully obvious reasons. 😅

**Reply by ericwbailey** [reply]
Cool cool, thanks for the insight. I'm also team "design for skimming readers", so no arguments here. Perspective on what content needs to be kept for historical context is also helpful, thank you.

**Comment by hamishwillee** [comment]
The two questions that most pop up associated with deprecated are:
- what should I be using instead
- why is this deprecated

It would be good if whatever we do here could answer either or both questions. The word "discouraged" does not answer these questions - it just tells you that you shouldn't be using it in an even less precise way.

Just an idea - maybe we could do some of this this with options to the deprecated header, allowing the generated content to be changed to incorporate a precise definition of the term. So for  `{{deprecated_header ("legacy")}}` to get something like 

> **Deprecated:** This is a legacy feature. It should not be used but cannot be removed from the specification (yet) because it is too widely used.
>
> Blah blah - the rest of the standard gumph

What I like this idea is that:
- Definitions are front and centre - you can state what chrome deprecated feature and a chrome legacy feature mean to us. We can differentiate legacy from deprecated etc.
- Definitions are always the same - if you use the string then you have the same definition.
- The default doesn't change - if you don't specify the type then it's just deprecated. That means we could roll this out slowly. We can also undo quickly by just changing the macro.
- We can answer the "why" question a lot more clearly without having to restrict ourselves to one word. 

Another similar idea is that you can specify a clarification to deprecated, but it must always be a spec definition. So below "legacy" is the word used to describe the API in the spec (would love an url to spec too, but let's keep this simple): `{{deprecated_header ("legacy")}}`. This feels restrictive - e.g. we'd only be able to capture what the spec means, not what MDN means.

> **Deprecated:** The specification defines this feature as: `legacy`.
>
> Blah, all the rest


Most recently this came up today in [Issue with "CSP: block-all-mixed-content": Deprecated without explanation or suggested replacement. #9889](https://github.com/mdn/content/issues/9889) (answered)

**Comment by erosman** [comment]
AFA terminology, there are not a lot of replacement choices for "deprecated" that can be easily understood.

While on the subject of "deprecated", I would also like to mention the following ...

- I have often come across an API/method/etc page MDN and after reading the entire page I found out that it was deprecated (or unsupported). I strongly think that an indication of such on TOP of the page would be greatly beneficial.
- In case of deprecated API/method/etc pages on MDN, alternative method is sometimes mentioned but not always. It would be great to have it where possible and where not, to mention no alternative is available.
- Some deprecated methods are marked as deprecated in the listing but not on pages e.g. [Document.execCommand()](https://developer.mozilla.org/docs/Web/API/Document/execCommand) vs [Interact with the clipboard](https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/Interact_with_the_clipboard)
